spopt/locate/p_median.py

```diff
@@ -533,3 +535,528 @@ def solve(self, solver: pulp.LpSolver, results: bool = True):
+            and facilities, by default "euclidean".
+        name : str, optional
+            The name of the problem, by default "k-nearest-p-median".
+
+        Returns
+        -------
+
+        spopt.locate.p_median.KNearestPMedian
+
+        Examples
+        --------
+        >>> from spopt.locate import KNearestPMedian
+        >>> import geopandas
+
+        Create the input data and attributes.
+
+        >>> k_list = np.array([1, 1])
+        >>> demand_data = {
+        ...    'ID': [1, 2],
+        ...    'geometry': [Point(0.5, 1), Point(1.5, 1)],
+        ...    'demand': [1, 1]}
+        >>> facility_data = {
+        ...    'ID': [101, 102],
+        ...    'geometry': [Point(1,1), Point(0, 2), Point(2, 0)],
+        ...    'capacity': [1, 1, 1]}
+        >>> gdf_demand = geopandas.GeoDataFrame(demand_data, crs='EPSG:4326')
+        >>> gdf_fac = geopandas.GeoDataFrame(facility_data, crs='EPSG:4326')
+
+        Create and solve a ``KNearestPMedian`` instance from the geodataframe.
+
+        >>> k_nearest_pmedian = KNearestPMedian.from_geodataframe(
+        ...     gdf_demand, gdf_fac,'geometry','geometry', weights_cols='demand',
+        ...     2, facility_capacity_col='capacity', k_list = k_list)
+        >>> k_nearest_pmedian = k_nearest_pmedian.solve(pulp.PULP_CBC_CMD(msg=False))
+
+        Get the facility-client associations.
+
+        >>> for fac, cli in enumerate(k_nearest_pmedian.fac2cli):
+        ...     print(f"facility {fac} serving {len(cli)} clients")
+        facility 0 serving 1 clients
+        facility 1 serving 1 clients
+        facility 2 serving 0 clients
+
+        Get the total and average weighted travel cost.
+
+        >>> round(k_nearest_pmedian.problem.objective.value(), 3)
+        1.618
+        >>> round(k_nearest_pmedian.mean_dist, 3)
+        0.809
+
+        Get the k list for the last iteration.
+        >>> print(k_nearest_pmedian.k_list)
+        [2, 1]
+
+        """
+
+        # check the crs of two geodataframes
+        if gdf_demand.crs is None:
+            raise ValueError("GeoDataFrame gdf_demand does not have a valid CRS.")
+        if gdf_fac.crs is None:
+            raise ValueError("GeoDataFrame gdf_facility does not have a valid CRS.")
+        if gdf_demand.crs != gdf_fac.crs:
+            raise ValueError(
+                "Geodataframes crs are different: "
+                f"gdf_demand-{gdf_demand.crs}, gdf_fac-{gdf_fac.crs}"
+            )
+
+        # create the array of coordinate of clients and facilities
+        dem = gdf_demand[demand_col]
+        fac = gdf_fac[facility_col]
+        dem_data = np.array([dem.x.to_numpy(), dem.y.to_numpy()]).T
+        fac_data = np.array([fac.x.to_numpy(), fac.y.to_numpy()]).T
+
+        # check the values of k_list
+        if k_list is None:
+            k_list = np.full(len(dem_data), np.minimum(len(fac_data), 5))
+        else:
+            if not (k_list <= len(fac_data)).all():
```

Comment:
Are we enforcing that `k_list` be a `numpy.array` (or can it be a `list`)?

* If the answer is "yes", we may consider updating the argument and attribute name to `k_array` or, more generally, `k_values`.
* If the answer is "no", we should have logic to check if `k_list` is actually a `numpy.array` and cast it as such if it is not. Line 959 will break if the type of `k_list` is an actual `list`.